### PR TITLE
[Darwin] Add optional concurrent execution to MTRAsyncWorkQueue

### PR DESCRIPTION
**Pull request by jtung-apple**
This adds an optional initializer to MTRAsyncWorkQueue to allow for concurrent execution of work items.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: src/darwin/Framework/CHIPTests/MTRAsyncWorkQueueTests.m [hunk 1]
```diff
@@ -491,4 +491,65 @@ - (void)testContextLoss
...
+    __block int afterSleepCounter = 0;
+    __auto_type sleep1ReadyHandler = ^(id context, NSInteger retryCount, MTRAsyncWorkCompletionBlock completion) {
+        os_unfair_lock_lock(&counterLock);
+        beforeSleepCounter++;
+        if (beforeSleepCounter == 3) {
+            [first3WorkItemsExecutedExpectation fulfill];
+        }
+        os_unfair_lock_unlock(&counterLock);
+        sleep(1);
+        afterSleepCounter++;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/MTRAsyncWorkQueueTests.m:511]: Do we not need to lock around the "after" counters?

File: src/darwin/Framework/CHIPTests/MTRAsyncWorkQueueTests.m [hunk 2]
```diff
@@ -491,4 +491,65 @@ - (void)testContextLoss
...
+
+    MTRAsyncWorkItem * workItem2 = [[MTRAsyncWorkItem alloc] initWithQueue:dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0)];
+    workItem2.readyHandler = sleep1ReadyHandler;
+    [workQueue enqueueWorkItem:workItem2 descriptionWithFormat:@"work item %d", 2];
+
+    MTRAsyncWorkItem * workItem3 = [[MTRAsyncWorkItem alloc] initWithQueue:dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0)];
+    workItem3.readyHandler = sleep1ReadyHandler;
+    [workQueue enqueueWorkItem:workItem3 descriptionWithFormat:@"work item %d", 3];
+
+    // This is the item after the first 3, and should onl execute when one of them finished
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/MTRAsyncWorkQueueTests.m:530]:
```suggestion
    // This is the item after the first 3, and should only execute when one of them finished
```

File: src/darwin/Framework/CHIPTests/MTRAsyncWorkQueueTests.m [hunk 3]
```diff
@@ -491,4 +491,65 @@ - (void)testContextLoss
...
+    XCTestExpectation * lastWorkItemWaitedExpectation = [self expectationWithDescription:@"Last work item waited properly"];
+    MTRAsyncWorkItem * workItemLast = [[MTRAsyncWorkItem alloc] initWithQueue:dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0)];
+    workItemLast.readyHandler = ^(id context, NSInteger retryCount, MTRAsyncWorkCompletionBlock completion) {
+        // expect this to have waited until at least one of the above items finished after sleep() and incremented counter
+        os_unfair_lock_lock(&counterLock);
+        if (afterSleepCounter > 0) {
+            [lastWorkItemWaitedExpectation fulfill];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/MTRAsyncWorkQueueTests.m:537]: Is there a reason this is not just `XCTAssertTrue(afterSleepCounter > 0)` followed by the fulfil?

**Review by bzbarsky-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

File: src/darwin/Framework/CHIPTests/MTRAsyncWorkQueueTests.m (hunk 1 above)
Comment (jtung-apple) [reply src/darwin/Framework/CHIPTests/MTRAsyncWorkQueueTests.m:512]: Done.

File: src/darwin/Framework/CHIPTests/MTRAsyncWorkQueueTests.m (hunk 3 above)
Comment (jtung-apple) [reply src/darwin/Framework/CHIPTests/MTRAsyncWorkQueueTests.m]: Done.

**Pull request merged**